### DISCUSSION TITLE
Optimizations and other refactor

### DISCUSSION THREAD

**Discussion by pdpino** [hoangnm/react-native-week-view#209] [Ideas]
## Problem

The component may run into performance issues when rendering too many days/weeks, and/or too many events. For example this happens with 500 evts:

https://user-images.githubusercontent.com/20805451/170147521-804d8550-92df-4518-9d92-19a76ab49406.mp4

* See the VirtualizedList warning
* There are dropped frames when scrolling
* Drag and dropping one event takes a lot of time (specifically, dropping the event is very slow)

## Possible causes

* When only one event is modified --> the `events` prop changes --> all `Event`s are re-rendered (e.g. when dragging an event)
* Each item of the VirtualizedList is too heavy/complex
* When scrolling horizontally, the page appending and prepending is done in place (I don't even remember why :sweat_smile:)
* When moving to a distant date (e.g. with `goToDate()`), many pages have to be added to the lists (see for example: https://github.com/hoangnm/react-native-week-view/issues/207)

(I've been using android systrace for profiling, but is still a limited tool)

## What now

I've been trying to improve these, not only for performance, but also for other features, for example:

1. The horizontal pinch-gesture (https://github.com/hoangnm/react-native-week-view/issues/130) will probably require a refactor to the VirtualizedList items. Implementation details: (skip if wanted)
    * Currently each item passed to the `VirtualizedList` represents one page of `numberOfDays` (e.g. 7 days)
    * This implies that when `numberOfDays` changes, the items must be recalculated and re-rendered from scratch to align items with pages (even though the events are the same). (See `componentDidUpdate()` [here](https://github.com/hoangnm/react-native-week-view/blob/master/src/WeekView/WeekView.js#L93))
    * A possible solution is to design each item to represent one day --> a change in `numberOfDays` would not require recalculating the items
2. Such refactor would allow fixing much more easily https://github.com/hoangnm/react-native-week-view/issues/128, https://github.com/hoangnm/react-native-week-view/issues/141#issuecomment-942321193, https://github.com/hoangnm/react-native-week-view/issues/185, https://github.com/hoangnm/react-native-week-view/issues/186
3. I want to create a first version of a `<MonthView>` component (https://github.com/hoangnm/react-native-week-view/issues/95) with these considerations in mind

---

@hoangnm any thoughts? I'll probably be submitting some PRs

cc: @rdewolff, @Salmankhan033, @Vidxyz

**Comment by greenboots88** [comment]
@pdpino @hoangnm the performance issues being experienced here and in #259 is due to the VirtualizedList. If you use your old library version from 2 years ago without VirtualizedList, the performance and speed is much better.

**Reply by pdpino** [reply]
@greenboots88 thanks, good to know!

Can you confirm which version/commit did you use? The VirtualizedList was introduced after 0.0.19
